lib/awful/client/focus.lua

```diff
@@ -0,0 +1,230 @@
+---------------------------------------------------------------------------
+--- Keep track of the focused clients.
+--
+-- @author Julien Danjou &lt;julien@danjou.info&gt;
+-- @copyright 2008 Julien Danjou
+-- @release @AWESOME_VERSION@
+-- @submodule client
+---------------------------------------------------------------------------
+local util = require("awful.util")
+
+local capi =
+{
+    screen = screen,
+    client = client,
+}
+
+-- We use a metatable to prevent circular dependency loops.
+local screen
+do
+    screen = setmetatable({}, {
+        __index = function(_, k)
+            screen = require("awful.screen")
+            return screen[k]
+        end,
+        __newindex = error -- Just to be sure in case anything ever does this
+    })
+end
+
+local client
+do
+    client = setmetatable({}, {
+        __index = function(_, k)
+            client = require("awful.client")
+            return client[k]
+        end,
+        __newindex = error -- Just to be sure in case anything ever does this
+    })
+end
+
+local focus = {history = {}}
+
+local internal = {}
+
+local function get_screen(s)
+    return s and capi.screen[s]
+end
+
+--- Remove a client from the focus history
+--
+-- @client c The client that must be removed.
+-- @function awful.client.focus.history.delete
+function focus.history.delete(c)
+    for k, v in ipairs(internal) do
+        if v == c then
+            table.remove(internal, k)
+            break
+        end
+    end
+end
+
+--- Focus a client by its relative index.
+--
+-- @function awful.client.focus.byidx
+-- @param i The index.
+-- @client[opt] c The client.
+function focus.byidx(i, c)
+    local target = client.next(i, c)
+    if target then
+        target:emit_signal("request::activate", "client.focus.byidx",
+                           {raise=true})
+    end
+end
+
+--- Filter out window that we do not want handled by focus.
+-- This usually means that desktop, dock and splash windows are
+-- not registered and cannot get focus.
+--
+-- @client c A client.
+-- @return The same client if it's ok, nil otherwise.
+-- @function awful.client.focus.filter
+function focus.filter(c)
+    if c.type == "desktop"
+        or c.type == "dock"
+        or c.type == "splash"
+        or not c.focusable then
+        return nil
+    end
+    return c
+end
+
+--- Update client focus history.
+--
+-- @client c The client that has been focused.
+-- @function awful.client.focus.history.add
+function focus.history.add(c)
+    -- Remove the client if its in stack
+    focus.history.delete(c)
+    -- Record the client has latest focused
+    table.insert(internal, 1, c)
+end
+
+-- Copy of awful.client.visible...
```

Comment:
Right, I will change this. I tried hard to avoid require("awful.client") dep loop, then I gave up because the code is a spaghetti ball. I tried to improve/fix/mitigate your complains about awful code. It isn't a total success, but at least it make awful.client smaller...